### PR TITLE
Add support for string and time indexed event stats 

### PR DESCRIPTION
**Pull request by MatMeredith**
Tests https://github.com/Metaswitch/cpp-common/pull/658

### CODE REVIEW

File: src/test_snmp.cpp [hunk 1]
```diff
@@ -12,6 +12,8 @@
 #endif
 
 #include "test_snmp.h"
+#include <string>
+
```
Comment (johadalin) [inline src/test_snmp.cpp]: blank line

File: src/test_snmp.cpp [hunk 2]
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
   delete tbl;
 }
 
+TEST_F(SNMPTest, LatencyScopeCalculations)
+{
+  cwtest_completely_control_time(true);
+
+  // Create a table.  Use a different OID base because otherwise the indices
+  // clash with later tests.
+  SNMP::EventAccumulatorByScopeTable* tbl = SNMP::EventAccumulatorByScopeTable::create("latency", test_oid);
```
Comment (johadalin) [inline src/test_snmp.cpp]: You say 'use a different OID base', but then go on to use the same `test_oid` as elsewhere. Am i missing something here?

File: src/test_snmp.cpp [hunk 3]
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
...
+
+  // Just put one sample in (which should have a variance of 0).  This should
+  // show up in only the current 5 minute stats.
+  tbl->accumulate(100);
+
+  // Check the counts
+  ASSERT_EQ(0, snmp_get(".1.2.2.1.7.1.4.110.111.100.101"));
```
Comment (johadalin) [inline src/test_snmp.cpp:102]: Can you add comments for the other two OIDs here as well? and those below. Seems really useful to give them human readable values

File: src/test_snmp.cpp [hunk 4]
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
...
+
+  // Average should be 400
+  ASSERT_EQ(400, snmp_get(".1.2.2.1.3.1.4.110.111.100.101"));
+  // HWM should be 500
+  ASSERT_EQ(500, snmp_get(".1.2.2.1.5.1.4.110.111.100.101"));
+  // LWM should be 300
+  ASSERT_EQ(300, snmp_get(".1.2.2.1.6.1.4.110.111.100.101"));
+  // Count should be 2
+  ASSERT_EQ(2, snmp_get(".1.2.2.1.7.1.4.110.111.100.101"));
+
```
Comment (johadalin) [inline src/test_snmp.cpp]: Why no test for variance here?

File: src/test_snmp.cpp [hunk 5]
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
...
+    {
+      ASSERT_EQ(0, snmp_get(".1.2.2.1." + std::to_string(stat) + "." + std::to_string(time_period) + ".4.110.111.100.101"));
+    }
+  }
+
+  cwtest_reset_time();
+  delete tbl;
+}
+
+enum EventStats
```
Comment (johadalin) [inline src/test_snmp.cpp:147]: Are we able to use these ENUM values in the test above to make it more readable?

File: src/test_snmp.cpp [hunk 6]
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
...
+  PREV_5S = 1,
+  CURR_5M = 2,
+  PREV_5M = 3
+};
+
+TEST_F(SNMPTest, StringBasedEventStats)
+{
+  cwtest_completely_control_time(true);
+
+  // Create a table.  Use a different OID base because otherwise the indices?// clash with later tests.
```
Comment (johadalin) [inline src/test_snmp.cpp]: Feel like this is a typo here. Also, as above, you're using the same `test_oid`.

File: src/test_snmp.cpp [hunk 7]
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
...
+
+  // Just put one sample in (which should have a variance of 0).  This should
+  // show up in only the current 5 minute stats.
+  std::string string1 = "server1.dc1.zone";
+  std::string string1_oid = "16.115.101.114.118.101.114.49.46.100.99.49.46.122.111.110.101";
+  std::string string2 = "server2.dc1.zone";
+  std::string string2_oid = "16.115.101.114.118.101.114.50.46.100.99.49.46.122.111.110.101";
+  tbl->accumulate(string1, 100);
+
+  //snmp_walk_debug(".1.2.2");
```
Comment (johadalin) [inline src/test_snmp.cpp]: If this is just for debugging what's going on, it should be removed

File: src/test_snmp.cpp (hunk 2 above)
Comment (johadalin) [reply src/test_snmp.cpp]: Just seen the commit comment saying no need to use different OIDs. Can you update the comments to reflect this?

File: src/test_snmp.cpp (hunk 7 above)
Comment (johadalin) [reply src/test_snmp.cpp]: Also, is this whole test function just for debugging? If so, shouldn't really be in the final code, right?

File: src/test_snmp.cpp [hunk 8]
```diff
@@ -86,6 +88,190 @@ TEST_F(SNMPTest, LatencyCalculations)
...
+
+  // Check that there are the right number of entries in the table (3 time
+  // periods * 1 string index * 5 metrics)
+  std::vector<std::string> entries = snmp_walk(".1.2.2");
```
Comment (johadalin) [inline src/test_snmp.cpp]: is .1.2.2 the test_oid value? can we use `test_oid` here instead?

**Review by johadalin — CHANGES_REQUESTED** [review]
The tests seem decent, but there are a number of places where the code just needs a clean up. Can you take another look over it to make sure comments are correct, and you've cleaned up the debug steps you put in.

File: src/test_snmp.cpp (hunk 7 above)
Comment (MatMeredith) [reply src/test_snmp.cpp]: The function is just for debugging, but it's not going to be the last time that I (or someone else working on this wants it).   This isn't production code, so is there a strong argument for removing it?

**Comment by MatMeredith** [comment]
@johadalin Thanks -- sorry -- this was a bit of a rushed check in.   I've addressed the most blatant issues that you've pointed out but there is a lot more that could be done in terms of making existing tests better.   The main thing I'd like to do is to generalise the time_string_event_oid function and use this throughout the other tests to make the OIDs we're testing much more readable as I have done with the new test.   

I'll come back to this if I get time tomorrow, otherwise I'm happy that the new test I've added raises the quality bar compared to what was already there, and that I've also made stuff better (e.g. adding the LatencyScopeCalculations test that was previously missing and making it more thorough than the LatencyCalculations test that it is copied from)!

Is that OK?

**Comment by johadalin** [comment]
Sounds like a reasonable plan. All looks good as is to me. Happy to leave generalising the time_string_event_oid function as a future enhancement, possibly tracked in an issue or somehting somewhere. Otherwise this looks good to merge now :+1:

**Review by johadalin — APPROVED** [review]

**Pull request merged**